File: code_snippets/ts/src/use_cases/microservices/subscription_object.ts
import * as restate from "@restatedev/restate-sdk/lambda";

// <start_here>
// <mark_2>
const subscriptionObject = restate.object({
  // </mark_2>
  name: "SubscriptionObject",
  handlers: {
    // <mark_2>
    add: async (ctx: restate.ObjectContext, req: SubscriptionRequest) => {
      // </mark_2>
      const paymentId = ctx.rand.uuidv4();

      // <mark_1>
      ctx.set("subscription", "awaiting_payment");
      // </mark_1>
      const success = await tryPaymentFiveTimes(ctx, req, paymentId);
      if (!success) {
        // <mark_1>
        ctx.set("subscription", "payment_failed");
        // </mark_1>
        return;
      }

      // <mark_1>
      ctx.set("subscription", "creating_subscription");
      // </mark_1>
      await ctx.run(() => createSubscription(req.userId, req.subscription));

      // <mark_1>
      ctx.set("subscription", "created");
      // </mark_1>
    },
  },
});

// <mark_3>
export const awsLambdaHandler = restate
  .endpoint()
  .bind(subscriptionObject)
  .handler();
// </mark_3>
// <end_here>

type SubscriptionRequest = {
  creditCard: string;
  userId: any;
  subscription: string;
};

function createRecurringPayment(creditCard: string, paymentId: string): string {
  return "";
}

function createSubscription(userId: string, subscription: string) {
  return undefined;
}

async function tryPaymentFiveTimes(
  ctx: restate.ObjectContext,
  req: SubscriptionRequest,
  paymentId: string
) {
  await ctx.run(
    "payment",
    () => createRecurringPayment(req.creditCard, paymentId),
    { maxRetryAttempts: 5 }
  );
  return true;
}

File: code_snippets/python/src/use_cases/microservices/subscription_object.py
import uuid

import restate
from pydantic import BaseModel
from restate import VirtualObject, ObjectContext


class SubscriptionRequest(BaseModel):
    user_id: str
    credit_card: str
    subscription: str


# <start_here>
subscription_object = VirtualObject("SubscriptionService")


# <mark_2>
@subscription_object.handler()
async def add(ctx: ObjectContext, req: SubscriptionRequest):
    # </mark_2>
    payment_id = await ctx.run("payment id", lambda: str(uuid.uuid4()))

    # <mark_1>
    ctx.set("subscription", "awaiting_payment")
    # </mark_1>
    success = await ctx.run(
        "recurring payment",
        lambda: create_recurring_payment(req.credit_card, payment_id),
    )
    if not success:
        # <mark_1>
        ctx.set("subscription", "payment_failed")
        # </mark_1>
        return

    # <mark_1>
    ctx.set("subscription", "creating_subscription")
    # </mark_1>
    await ctx.run("subscription", lambda: create_subscription(req.user_id, req.subscription))
    # <mark_1>
    ctx.set("subscription", "created")
    # </mark_1>


# <mark_3>
aws_lambda_handler = restate.app([subscription_object])
# </mark_3>
# <end_here>


def create_recurring_payment(credit_card, payment_id):
    return None


def create_subscription(user_id, subscription):
    return None
